### DISCUSSION TITLE
Can we change struct `SCColor` to class?

### DISCUSSION THREAD

**Discussion by huchim** [ShapeCrawler/ShapeCrawler#511] [Ideas]
Hi @adamshakhabov!...

I was thinkig about the posibility to set highlight color using the `accent` color (or other), but with the current implementation I can only use RGB...

Method [`ToRgbColorModelHex(this SCColor color)`](https://github.com/ShapeCrawler/ShapeCrawler/blob/cb5cad3f70e6ffc22243d9a2080fa5f021796f1b/src/ShapeCrawler/Extensions/RunPropertiesExtensions.cs#L41) could be replace with a "color factory", for example `ISolidColorFactory.Create(color)` in order to create `TypedOpenXmlCompositeElement` of type "`a:srgbClr`", "`a:hslClr`", "`a:sysClr`", "`a:schemeClr`", etc...

```cs
// base class
public class SCColor
{
    public static readonly SCColor Black = new(0, 0, 0);
}

// extends other
public class SCSchemeColor : SCColor
{
  public static readonly SCSchemeColor Accent = new("accent");
  public SCSchemeColor(string name) : base(0, 0, 0, 0) { /** impl **/}
}

// a color factory
public interface ISolidColorFactory {
  // create
  public TypedOpenXmlCompositeElement Create(SCColor color); // RGB default
  public TypedOpenXmlCompositeElement Create(SCSchemeColor color);
  public TypedOpenXmlCompositeElement Create(SCHslColor color);
  public TypedOpenXmlCompositeElement Create(SCPresetColor color);
  public TypedOpenXmlCompositeElement Create(SCRgbColor color);
  public TypedOpenXmlCompositeElement Create(SCSystemColor color);

  // parse
  public SCColor  GetColor(TypedOpenXmlCompositeElement color);
}
```

I'm not sure whether `OneOf<SCColor, SCSchemeColor>`  can be used...

This just a question, struct if the correct type for the requirement, but this could be other option.

Have a nice day

**Comment by AdamShakhabov** [comment]
Hello @huchim,

From a user's perspective, the idea is to be able to set, for example, a Theme Color for a portion of a paragraph, correct?

![image](https://github.com/ShapeCrawler/ShapeCrawler/assets/5969304/a5bb8d28-e29b-4fc3-be42-da567f7b7503)

**Reply by huchim** [reply]
Yes, that is the idea, but may be not only on paragraphs...

**Reply by AdamShakhabov** [reply]
Yes, we can use the implementation that you suggested above, but for the high-level APIs, I had a plan to extend `IColorFormat` like:

```c#
public interface IColorFormat
{
    //...

    /// <summary>
    ///     Gets or sets theme schema color.
    /// </summary>
    SCSchemaColor SchemaColor { set; get; }
    
    /// <summary>
    ///     Gets or sets predefined system color.
    /// </summary>
    SCSystemColor SystemColor { set; get; }
}

public enum SCSchemaColor
{
    NotDefined,
    Background1,
    Text1,
    Accent1,
    //...
}

public enum SCSystemColor
{
    NotDeifned,
    ScrollBar,
    Background,
    //...
}
```

**Reply by huchim** [reply]
I have a few questions

In this case, how could I know what color is the correct?

```cs
node.ColorFormat.RgbColor = RGBColor.White; // Custom color? 
node.ColorFormat.HslColor = SCHslColor.Red; // Custom color?
node.ColorFormat.SchemaColor = SCSchemaColor.Text1; 
node.ColorFormat.SystemColor= SCSystemColor.ScrollBar; 
```

I can see you are traversing nodes until find the color type and the equivalent RGB color (`(SCColorType, string)`).

I found implementations to get these colors:

- `schemeClr` (Scheme Color)
- `prstClr` (Preset Color)
- `srgbClr` (RGB Color Model - Hex Variant)
- `sysClr` (System Color)
- *Without implementation?*
  - `hslClr` (Hue, Saturation, Luminance Color Model)
  - `scrgbClr` (RGB Color Model - Percentage Variant)

I'm sorry if I misunderstanding the problem, my English is not good :grimacing:, in my opinion, an abstract factory in `ColorFormat` and a sealed `SCSystemColor` [class enum](https://learn.microsoft.com/en-us/dotnet/architecture/microservices/microservice-ddd-cqrs-patterns/enumeration-classes-over-enum-types) could be better.

Thanks your your time, I'm able to contribute to your project as you want :muscle:

**Reply by AdamShakhabov** [reply]
Carlos, thanks for your hard work!👍

> how could I know what color is the correct?

For example, when a user sets `IColorFormat.SchemaColor`, then `IColorFormat.SystemColor` and others should be set to `NotDefined`. Plus, there's already a property `IColorFormat.ColorType` that makes it easier to define which color type is being used.

> Without implementation?

You're right. I was only focused on solid type colors at that time. We should also cover `hslClr` and `scrgbClr`. We're open to ideas, so feel to share your suggestions.

As for using Enumeration classes over Enum, I'll be honest, I don't have much experience with this. I agree that the built-in Enum has its limitations. From what I've quick researched, this wrapper is useful in _applications_. But why not? I think we can try this in our _library_ type solution🙂

**Reply by huchim** [reply]
I was digging in the OpenXML library and I found that they uses Enum with an attribute `EnumStringAttribute` and the `EnumValue` wrapper (there is no `EnumInfoLookup` in the version 3.0).

```cs
enum SCSchemaColor {
 [EnumString("accent1")]
 Accent1
}

// 
var classEnum = EnumValue<SCSchemaColor>(SCSchemaColor.Accent1);

// EnumValue works as a factory in the version 3.0
var accent1 = EnumValue<SCSchemaColor>.Create("accent1);
```

:100: I know you know that, but following your idea, the setter `SystemColor` could reset to `NotDefined` other types, 

```cs
node.ColorFormat.RgbColor = RGBColor.White; // Maybe we can use the method `SetRgbByHex()`  instead this
node.ColorFormat.HslColor = SCHslColor.Red; // Reset other color types 
node.ColorFormat.SchemaColor = SCSchemaColor.Text1; // Reset other color types 
node.ColorFormat.SystemColor= SCSystemColor.ScrollBar; // The last is the winner!
```

A `SCColor` is would be required if colortype is `RgbColor` and must throw an exception in somepoint...

:crossed_fingers: Let me create a new PR to refactor your color getter to abstract factory without create new enums or classes. 

:v::v:
